core/src/main/java/org/apache/iceberg/GenericTrackedFile.java

```diff
@@ -0,0 +1,773 @@
+  // All fields are made optional to support different content types
+  static final Types.StructType BASE_TYPE =
+      Types.StructType.of(
+          TrackedFile.CONTENT_TYPE,
+          TrackedFile.LOCATION,
+          TrackedFile.FILE_FORMAT,
+          TrackedFile.PARTITION_SPEC_ID,
+          TrackedFile.SORT_ORDER_ID,
+          TrackedFile.RECORD_COUNT,
+          TrackedFile.FILE_SIZE_IN_BYTES,
+          TrackedFile.KEY_METADATA,
+          TrackedFile.SPLIT_OFFSETS,
+          TrackedFile.EQUALITY_IDS,
+          TrackedFile.REFERENCED_FILE,
+          // Tracking info fields
+          TrackingInfo.STATUS.asOptional(),
+          TrackingInfo.SNAPSHOT_ID,
+          TrackingInfo.SEQUENCE_NUMBER,
+          TrackingInfo.FILE_SEQUENCE_NUMBER,
+          TrackingInfo.FIRST_ROW_ID,
+          // Deletion vector fields
+          DeletionVector.OFFSET,
+          DeletionVector.SIZE_IN_BYTES,
+          DeletionVector.INLINE_CONTENT,
+          // Manifest stats fields
+          ManifestStats.ADDED_FILES_COUNT.asOptional(),
+          ManifestStats.EXISTING_FILES_COUNT.asOptional(),
+          ManifestStats.DELETED_FILES_COUNT.asOptional(),
+          ManifestStats.ADDED_ROWS_COUNT.asOptional(),
+          ManifestStats.EXISTING_ROWS_COUNT.asOptional(),
+          ManifestStats.DELETED_ROWS_COUNT.asOptional(),
+          ManifestStats.MIN_SEQUENCE_NUMBER.asOptional());
+
+  /** Used by Avro reflection to instantiate this class when reading manifest files. */
+  public GenericTrackedFile(Schema avroSchema) {
+    this(AvroSchemaUtil.convert(avroSchema).asStructType());
+    this.avroSchema = avroSchema;
+  }
+
+  /** Used by internal readers to instantiate this class with a projection schema. */
+  GenericTrackedFile(Types.StructType projection) {
+    super(BASE_TYPE, projection);
+    this.avroSchema = AvroSchemaUtil.convert(projection, "tracked_file");
+  }
+
+  public GenericTrackedFile() {
+    super(BASE_TYPE.fields().size());
+  }
+
+  /**
+   * Full constructor for creating a tracked file.
+   *
+   * @param contentType the type of content (DATA, POSITION_DELETES, etc.)
+   * @param location the file location
+   * @param fileFormat the file format
+   * @param partitionSpecId the partition spec ID
+   * @param recordCount the number of records
+   */
+  public GenericTrackedFile(
+      FileContent contentType,
+      String location,
+      FileFormat fileFormat,
+      int partitionSpecId,
+      long recordCount,
+      Long fileSizeInBytes) {
+    super(BASE_TYPE.fields().size());
+    this.contentType = contentType;
+    this.location = location;
+    this.fileFormat = fileFormat;
+    this.partitionSpecId = partitionSpecId;
+    this.recordCount = recordCount;
+    this.fileSizeInBytes = fileSizeInBytes;
+  }
+
+  /**
+   * Copy constructor.
+   *
+   * @param toCopy a tracked file to copy
+   * @param copyStats whether to copy stats
+   */
+  private GenericTrackedFile(GenericTrackedFile toCopy, boolean copyStats) {
```

Comment:
For v4, we can remove the paths that use a `copyStats` boolean. Instead, we should use a set of field IDs to identify the fields to copy, much like the newer copy API.